test/end2end_test.go

```diff
@@ -5509,6 +5510,294 @@ func (s) TestClientForwardsGrpcAcceptEncodingHeader(t *testing.T) {
+	})
+
+	t.Run("stream", func(t *testing.T) {
+		testStreamSetSendCompressorFailure(t, resCompressor, wantErr)
+	})
+}
+
+func (s) TestUnadvertisedSetSendCompressorFailure(t *testing.T) {
+	// Disable client compressor advertisement.
+	defer func(b bool) { envconfig.AdvertiseCompressors = b }(envconfig.AdvertiseCompressors)
+	envconfig.AdvertiseCompressors = false
+
+	resCompressor := "gzip"
+	wantErr := status.Error(codes.Unknown, "unable to set send compressor: client does not support compressor \"gzip\"")
+
+	t.Run("unary", func(t *testing.T) {
+		testUnarySetSendCompressorFailure(t, resCompressor, wantErr)
+	})
+
+	t.Run("stream", func(t *testing.T) {
+		testStreamSetSendCompressorFailure(t, resCompressor, wantErr)
+	})
+}
+
+func testUnarySetSendCompressorFailure(t *testing.T, resCompressor string, wantErr error) {
+	ss := &stubserver.StubServer{
+		EmptyCallF: func(ctx context.Context, in *testpb.Empty) (*testpb.Empty, error) {
+			if err := grpc.SetSendCompressor(ctx, resCompressor); err != nil {
+				return nil, err
+			}
+			return &testpb.Empty{}, nil
+		},
+	}
+	if err := ss.Start(nil); err != nil {
+		t.Fatalf("Error starting endpoint server: %v", err)
+	}
+	defer ss.Stop()
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	if _, err := ss.Client.EmptyCall(ctx, &testpb.Empty{}); !equalError(err, wantErr) {
+		t.Fatalf("Unexpected unary call error, got: %v, want: %v", err, wantErr)
+	}
+}
+
+func testStreamSetSendCompressorFailure(t *testing.T, resCompressor string, wantErr error) {
+	ss := &stubserver.StubServer{
+		FullDuplexCallF: func(stream testpb.TestService_FullDuplexCallServer) error {
+			if _, err := stream.Recv(); err != nil {
+				return err
+			}
+
+			if err := grpc.SetSendCompressor(stream.Context(), resCompressor); err != nil {
+				return err
+			}
+
+			return stream.Send(&testpb.StreamingOutputCallResponse{})
+		},
+	}
+	if err := ss.Start(nil); err != nil {
+		t.Fatalf("Error starting endpoint server: %v, want: nil", err)
+	}
+	defer ss.Stop()
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	s, err := ss.Client.FullDuplexCall(ctx)
+	if err != nil {
+		t.Fatalf("Unexpected full duplex call error, got: %v, want: nil", err)
+	}
+
+	if err := s.Send(&testpb.StreamingOutputCallRequest{}); err != nil {
+		t.Fatalf("Unexpected full duplex call send error, got: %v, want: nil", err)
+	}
+
+	if _, err := s.Recv(); !equalError(err, wantErr) {
+		t.Fatalf("Unexpected full duplex recv error, got: %v, want: nil", err)
+	}
+}
+
+func (s) TestUnarySetSendCompressorAfterHeaderSendFailure(t *testing.T) {
+	ss := &stubserver.StubServer{
+		EmptyCallF: func(ctx context.Context, in *testpb.Empty) (*testpb.Empty, error) {
+			// Send headers early and then set send compressor.
+			grpc.SendHeader(ctx, metadata.MD{})
+			err := grpc.SetSendCompressor(ctx, "gzip")
+			if err == nil {
+				t.Error("Wanted set send compressor error")
```

Comment:
Let's return `&testpb.Empty{}, nil` here, otherwise I believe the below (`return nil, <err=nil>`) will panic(?) because of the nil message.  (Similar elsewhere in this PR.)